File: server/Services/VaultsService.cs
namespace keepr.Services;

public class VaultsService
{
    private readonly VaultsRepository _vaultsRepository;

    public VaultsService(VaultsRepository vaultsRepository)
    {
        _vaultsRepository = vaultsRepository;
    }

    internal Vault CreateVault(VaultCreationDTO creationData, string userId)
    {
        return _vaultsRepository.CreateVault(creationData, userId);
    }

    private Vault GetVaultById(int vaultId)
    {
        Vault vault = _vaultsRepository.GetVaultById(vaultId);
        if (vault == null) throw new Exception($"invalid Id: {vaultId}");

        return vault;
    }

    internal Vault GetVaultById(int vaultId, string userId)
    {
        Vault vault = _vaultsRepository.GetVaultById(vaultId);
        if (vault.IsPrivate == true && vault.CreatorId != userId) throw new Exception($"Invalid id, vault is private");
        return vault;
    }

    internal Vault UpdateVault(int vaultId, string userId, VaultCreationDTO creationData)
    {
        Vault vault = GetVaultById(vaultId);

        if (vault.CreatorId != userId) throw new Exception($"invalid id: {vaultId}");

        vault.Name = creationData.Name ?? vault.Name;
        vault.IsPrivate = creationData.IsPrivate ?? vault.IsPrivate;

        _vaultsRepository.UpdateVault(vault);

        return vault;
    }

    internal Vault DeleteVault(int vaultId, string userId)
    {
        Vault vault = GetVaultById(vaultId);

        if (vault.CreatorId != userId) throw new Exception();

        _vaultsRepository.DeleteVault(vaultId);

        return vault;
    }

    internal List<Vault> GetMyVault(string userId)
    {
        List<Vault> vaults = _vaultsRepository.GetMyVault(userId);
        return vaults;
    }

    internal List<Vault> GetUserVault(string profileId)
    {
        List<Vault> vaults = _vaultsRepository.GetUserVault(profileId);

        return vaults.FindAll(vault => vault.IsPrivate == false || vault.CreatorId != profileId);
    }
}

File: client/src/services/VaultsService.js
import { logger } from "@/utils/Logger.js";
import { api } from "./AxiosService.js";
import { AppState } from "@/AppState.js";
import { Vault } from "@/models/Vault.js";

class VaultsService
{
    async deleteVault(vaultId) {
       const response = await api.delete(`api/vaults/${vaultId}`);
       logger.log('delete vault', response.data)
       const vaultToDelete = AppState.vaults.findIndex((vault) => vault.id == vaultId)
       AppState.vaults.splice(vaultToDelete,1)
    }

    async updateVault(vaultId, vaultData) {
      const response = await api.put(`api/vaults/${vaultId}`,vaultData)
      logger.log('update vault ', response.data)
      const newVault = new Vault(response.data)
      AppState.vaults.push(newVault)
    }

    async getMyVault() {
      const response = await api.get('account/vaults')
      logger.log('get my vaults', response.data)
      AppState.vaults = response.data.map(vault => new Vault(vault))
    }

    async getUserVault(profileId) {
      AppState.vaults = null
      const response = await api.get(`api/profiles/${profileId}/vaults`)
      logger.log('get user vaults', response.data)
      AppState.vaults = response.data.map(vault => new Vault(vault))
    }

    async getVaultById(vaultId) 
    {
      AppState.vault = null
      const response = await api.get(`api/vaults/${vaultId}`)
      logger.log('get vault by id', response.data)
      AppState.vault = new Vault(response.data);
    }

    async createVault(vaultData)
    {
      const response = await api.post("api/vaults", vaultData)
      logger.log('create vaults', response.data)
      const newVault = new Vault(response.data)
      AppState.vaults.push(newVault)
    }
}

export const vaultsService = new VaultsService();